tests/end_to_end_cases/write_api.rs

```diff
@@ -425,3 +430,183 @@ async fn test_write_routed_errors() {
+    let mut router_mgmt = router.management_client();
+    router_mgmt
+        .update_server_id(TEST_ROUTER_ID)
+        .await
+        .expect("set ID failed");
+
+    let target_1 = ServerFixture::create_single_use().await;
+    let mut target_1_mgmt = target_1.management_client();
+    target_1_mgmt
+        .update_server_id(TEST_TARGET_ID_1)
+        .await
+        .expect("set ID failed");
+
+    router_mgmt
+        .update_remote(TEST_REMOTE_ID_1, target_1.grpc_base())
+        .await
+        .expect("set remote failed");
+
+    let target_2 = ServerFixture::create_single_use().await;
+    let mut target_2_mgmt = target_2.management_client();
+    target_2_mgmt
+        .update_server_id(TEST_TARGET_ID_2)
+        .await
+        .expect("set ID failed");
+
+    router_mgmt
+        .update_remote(TEST_REMOTE_ID_2, target_2.grpc_base())
+        .await
+        .expect("set remote failed");
+
+    let target_3 = ServerFixture::create_single_use().await;
+    let mut target_3_mgmt = target_3.management_client();
+    target_3_mgmt
+        .update_server_id(TEST_TARGET_ID_3)
+        .await
+        .expect("set ID failed");
+
+    router_mgmt
+        .update_remote(TEST_REMOTE_ID_3, target_3.grpc_base())
+        .await
+        .expect("set remote failed");
+
+    let db_name_1 = rand_name();
+    let db_name_2 = rand_name();
+    for &db_name in &[&db_name_1, &db_name_2] {
+        create_readable_database(db_name, router.grpc_channel()).await;
+        create_readable_database(db_name, target_1.grpc_channel()).await;
+        create_readable_database(db_name, target_2.grpc_channel()).await;
+        create_readable_database(db_name, target_3.grpc_channel()).await;
+    }
+
+    // Set routing rules on the router:
+    for (db_name, remote_id) in &[
+        (db_name_1.clone(), TEST_REMOTE_ID_1),
+        (db_name_2.clone(), TEST_REMOTE_ID_2),
+    ] {
+        let mut router_db_rules = router_mgmt
+            .get_database(db_name)
+            .await
+            .expect("cannot get database on router");
+        let routing_config = RoutingConfig {
+            target: Some(NodeGroup {
+                nodes: vec![Node { id: *remote_id }],
+            }),
+        };
+        router_db_rules.routing_rules = Some(RoutingRules::RoutingConfig(routing_config));
+
+        router_mgmt
+            .update_database(router_db_rules)
+            .await
+            .expect("cannot update router db rules");
+    }
+
+    // Write some data
+    let line_1 = "cpu bar=1 100";
+    let line_2 = "disk bar=2 100";
+
+    let mut write_client = router.write_client();
+
+    for (&ref db_name, &ref line) in &[(&db_name_1, line_1), (&db_name_2, line_2)] {
+        let num_lines_written = write_client
+            .write(db_name, line)
+            .await
+            .expect("cannot write");
+        assert_eq!(num_lines_written, 1);
+    }
+
+    // The router will have split the write request by table name into two shards.
```

Comment:
I think it is routing based on database name, not table name

```suggestion
    // The router will have split the write request by database name.
```

Reply:
copy pasta is worst pasta